ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -131,6 +145,38 @@ private[platform] final class EventsBuffer[O: Ordering, E](
         }
       },
     )
+
+  private def indexAfter(bufferStartInclusiveSearchResult: SearchResult) =
+    bufferStartInclusiveSearchResult match {
+      case InsertionPoint(insertionPoint) => insertionPoint
+      case Found(foundIndex) => foundIndex + 1
+    }
+
+  private def filterAndChunkSlice[FILTER_RESULT](
+      bufferSlice: Vector[(Offset, ENTRY)],
+      filter: ENTRY => Option[FILTER_RESULT],
+  ) = {
+    var takenFiltered = 0
+    var lastScannedOffset = Offset.beforeBegin
+
+    val it = bufferSlice.iterator
+    val resultBuilder = Vector.newBuilder[(Offset, FILTER_RESULT)]
+
+    while (it.hasNext && takenFiltered < maxBufferedChunkSize) {
```

Comment:
It was a conscious decision for one reason: checking whether we need to offer a new chunk (see line 177 with it.hasNext). I'd love making this functional, but couldn't come up with an efficient solution due to:
* This function should compute a filtered collection of size `maxBufferedChunkSize`
* It should also return a `startExclusive` needed for the next chunk processing IF the underlying window has more elements to-be-scanned.

Reply:
Improved somehow the implementation by making it more concise and removing the mutable vars. Although it's shorter now, I am now concerned a bit about how obvious is to everyone why the last `.hasNext` correctly flags that there needs to be another fetched chunk. Let me know what do you think

Reply:
I like it a lot :clap: 
Re-using iterators is sometimes tricky, and generally discouraged in Scala, so I would suggest using splitAt, where you can covert _1 to Vector, and you can call the hasNext on _2

Reply:
If you use splitAt instead of take and call hasnext on _2, you are potentially evaluating a lot of filters (redundant work) until you reach a next Some(_) that yields true.

Reply:
Oh yeah, that's right.
It makes me wonder whether we need this option here at all (there is already a signalling stream end for the unfold one level up, can we just make that smarter, and make this simpler)?
Similary to the lastFilteredChunk: there we could just do always the head/tail split for chunkSize+1, which would be even more efficient, since it would make the range for DB query shorter -- and similarly simplifying this function.

Reply:
As discussed offline improvements here will come later (maybe a TODO)